fedimint-core/src/db/mod.rs

```diff
@@ -354,674 +400,762 @@ impl Database {
     }
 }
 
-#[apply(async_trait_maybe_send!)]
-pub trait IDatabaseTransactionOps<'a>: 'a + MaybeSend {
-    async fn raw_insert_bytes(&mut self, key: &[u8], value: &[u8]) -> Result<Option<Vec<u8>>>;
-
-    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>>;
-
-    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>>;
-
-    /// Returns an stream of key-value pairs with keys that start with
-    /// `key_prefix`. No particular ordering is guaranteed.
-    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> Result<PrefixStream<'_>>;
-
-    /// Same as [`Self::raw_find_by_prefix`] but the order is descending by key.
-    async fn raw_find_by_prefix_sorted_descending(
-        &mut self,
-        key_prefix: &[u8],
-    ) -> Result<PrefixStream<'_>>;
-
-    /// Delete keys matching prefix
-    async fn raw_remove_by_prefix(&mut self, key_prefix: &[u8]) -> Result<()>;
+fn module_instance_id_to_byte_prefix(module_instance_id: u16) -> Vec<u8> {
+    let mut prefix = vec![MODULE_GLOBAL_PREFIX];
+    module_instance_id
+        .consensus_encode(&mut prefix)
+        .expect("Error encoding module instance id as prefix");
+    prefix
+}
 
-    async fn rollback_tx_to_savepoint(&mut self) -> Result<()>;
+/// A database that wraps an `inner` one and adds a prefix to all operations,
+/// effectively creating an isolated partition.
+#[derive(Clone, Debug)]
+pub struct PrefixDatabase<Inner>
+where
+    Inner: Debug,
+{
+    prefix: Vec<u8>,
+    inner: Inner,
+}
 
-    /// Create a savepoint during the transaction that can be rolled back to
-    /// using rollback_tx_to_savepoint. Rolling back to the savepoint will
-    /// atomically remove the writes that were applied since the savepoint
-    /// was created.
-    ///
-    /// Warning: Avoid using this in fedimint client code as not all database
-    /// transaction implementations will support setting a savepoint during
-    /// a transaction.
-    async fn set_tx_savepoint(&mut self) -> Result<()>;
+impl<Inner> PrefixDatabase<Inner>
+where
+    Inner: Debug,
+{
+    // TODO: we should optimize these concatenations, maybe by having an internal
+    // `key: &[&[u8]]` that we flatten once, when passing to lowest layer, or
+    // something
+    fn get_full_key(&self, key: &[u8]) -> Vec<u8> {
+        let mut full_key = self.prefix.clone();
+        full_key.extend_from_slice(key);
+        full_key
+    }
 }
 
-/// Fedimint requires that the database implementation implement Snapshot
-/// Isolation. Snapshot Isolation is a database isolation level that guarantees
-/// consistent reads from the time that the snapshot was created (at transaction
-/// creation time). Transactions with Snapshot Isolation level will only commit
-/// if there has been no write to the modified keys since the snapshot (i.e.
-/// write-write conflicts are prevented).
-///
-/// Specifically, Fedimint expects the database implementation to prevent the
```

Comment:
Did this comment just get moved? This might still be helpful for understanding the concurrency semantics that Fedimint expects

Reply:
Ah I see it now